### ISSUE TITLE
Custom drawing functionality?

### ISSUE THREAD

**Issue by iliyang** [swistakm/pyimgui#49]
Hi, is there a way to custom-draw a region, like it is possible in the C++ `imgui`? Likely not, because I don't see `ImDrawList` exposed in the Python API, but is this planned?

**Comment by swistakm** [comment]
I'm not sure I understand your question. What exactly do you want to achieve? Could you elaborate or give a more specific example?

If you want to implement your own rendering then the `ImDrawList` objects are available indirectly in render callbacks (via `draw_data argument`).

**Comment by iliyang** [comment]
Sorry I wasn't clear the first time. What I want is to create my own widget that will have some area that will be custom-drawn. I want to be able to draw my own charts, histograms, etc using 2D primitives like lines, curves, ellipses, rectangles -- this sort of stuff. Is that currently possible with **pyimgui**? It would be truly great if yes, but how?

I am a complete **imgui** beginner, but I remember seeing some custom-draw C++ code that was adding primitives to `ImDrawList`. That's why I mentioned it. Please forgive my ignorance!

**Comment by iliyang** [comment]
Just to be clear, my entire application is an imgui window - everything that I display on screen are imgui controls. This is slightly different to traditional use cases where imgui supplements the user's application. What I'm doing is essentially using imgui instead of Qt or wxWidgets for a UI application because I need better performance.

I want to have my own controls that draw themselves. For example, I want to have a scrollable list where each item is a widget that plots some data (a histogram, function plot, etc) thay is not supported by imgui out of the box.

**Comment by swistakm** [comment]
Oh, now I understand. Thanks for clarification :)

Such thing is not possible right now but it is definitely doable. We will heave to expose some internal functions that return the current window object and also to create mapping for `ImGuiWindow` and extend the existing mapping for ImDrawList that will allow to manually add new drawing primitives.

I will try to take a look into this in my free time soon to see how it could be done. I'm sure I won't have time to ensure 100% feature compatibility on my own but I will try to at least set some foundation for others so new functions could be easily contributed.

**Comment by iliyang** [comment]
Thanks for the reply, I very much appreciate your time! The ability to take advantage of imgui's 2D drawing subsystem would be huge on its own!

In the mean time I hope option 1 from [this post](https://github.com/ocornut/imgui/issues/984#issuecomment-273241640) would be possible. That is, render my custom stuff to texture and then put this texture in a imgui widget. This is a bit less convenient as it has overhead (if I want to have many such widgets) and I'll have to use another drawing library. But I'd be fine with it if it's possible. Does **pyimgui** support placing images in widgets? A quick search returns [this discussion](https://github.com/ocornut/imgui/issues/122) on the imgui issue tracker makes me worry it's not currently possible with **pyimgui**...

**Comment by swistakm** [comment]
Yes, for now you can render directly to texture and display it as `imgui.core.image()` widget.

**Comment by iliyang** [comment]
Great, I will try and report, thanks!!

**Comment by lubieowoce** [comment]
Note for anyone encountering this issue: The [1.0.0a2](https://github.com/swistakm/pyimgui/releases/tag/1.0.0a2) pre-release has some `ImDrawList` functionality exposed – we can draw lines and [filled] rectangles, and adding bindings for the rest should be easy.

**Comment by vstepchik** [comment]
> Note for anyone encountering this issue: The [1.0.0a2](https://github.com/swistakm/pyimgui/releases/tag/1.0.0a2) pre-release has some `ImDrawList` functionality exposed – we can draw lines and [filled] rectangles, and adding bindings for the rest should be easy.

Already found the ImDrawList exposed, however, I'm missing the `ImDrawList::AddText` API. Are the missing parts to be added anytime soon?

**Comment by swistakm** [comment]
Released as `1.0.0`
